thewalrus/internal_modes/prepare_cov.py

```diff
@@ -0,0 +1,304 @@
+    inds = np.arange(lambd.shape[0])[lambd > thr]
+    lambd = lambd[inds]
+    R = lambd.shape[0]
+    eps = []
+    W = []
+    for j in range(M):
+        Rtemp = np.eye(R, dtype=np.complex128)
+        for l in range(R):
+            for m in range(R):
+                Rtemp[l, m] = np.sum(
+                    np.array(
+                        [
+                            V[indexconversion(j, k, njs), l].conj()
+                            * V[indexconversion(j, k, njs), m].conj()
+                            * np.sqrt(lambd[l])
+                            * np.sqrt(lambd[m])
+                            for k in range(njs[j])
+                        ]
+                    )
+                )
+        eps_temp, WT_temp = takagi(Rtemp)
+        signs = np.sign(WT_temp.real)[0]
+        for j, s in enumerate(signs):
+            if np.allclose(s, 0):
+                signs[j] = 1
+        WT_temp /= signs
+        eps.append(eps_temp)
+        W.append(np.real_if_close(WT_temp).T)
+    if F is not None:
+        chis = []
+        for k in range(R):
+            chi = np.sum(
+                np.array(
+                    [np.sqrt(rs[j]) * Fs[j] * V[j, k] / np.sqrt(lambd[k]) for j in range(len(Fs))]
+                ),
+                axis=0,
+            )
+            chis.append(chi / np.linalg.norm(chi))
+        return chis, eps, W
+    return eps, W
+
+
+def state_prep(eps, W, thresh=1e-4, hbar=2):
+    r"""
+    Computes the total covariance matrix (assuming zero displacement) of the initial state as determined by orthonormalization parameters.
+    Modes are ordered first by orthonormalization modes, then spatial modes, i.e. for R orthonormalization modes the 1st R modes of the
+    covariance matrix are for the first spatial mode, the next R lot of modes for the 2nd spatial mode, etc. If an orthonormal mode has
+    a fidelity with vacuum of 1 - thresh or higher then it is traced over, unless every mode is like that then the whole system is returned.
+
+    Args:
+        eps (list[array]): list of arrays of squeezing parameters for each spatial mode as determined by the orthonormalization procedure
+        W (list[array]): list of arrays of the orthonormalization amplitude matrix, one for each spatial mode
+
+    Returns:
+        (array): covariance matrix of the output state
+    """
+    if not np.allclose(len(eps), len(W)):
+        raise ValueError("len of eps must be equal to len of W")
+    epsBig = np.array([])
+    for i in eps:
+        epsBig = np.append(epsBig, i)
+
+    M = len(eps)
+    R = eps[0].shape[0]
+
+    Qvac = (hbar / 2) * np.eye(2 * M * R)
+    S = squeezing(epsBig)
+    Qinit = S @ Qvac @ S.T
+
+    Wbig = np.eye(M * R, dtype=np.complex128)
+    for j in range(M):
+        Wbig[R * j : R * (j + 1), R * j : R * (j + 1)] = W[j]
+
+    Q = interferometer(Wbig.T.conj()) @ Qinit @ interferometer(Wbig.T.conj()).T
+    Qswap = interferometer(swap_matrix(M, R)) @ Q @ interferometer(swap_matrix(M, R)).T
```

Comment:
All the Q matrices here are not the same as the Q matrix from `Qmat` to my knowledge. The notation might be a bit confusing.

Reply:
The other option is to call them all cov (or similar), think it's worth the effort?